Analyzer: Prefer .Length/Count/IsEmpty over Any()

CollinAlpert opened the pull request:
<!--

Make sure you have read the contribution guidelines: 
- https://docs.microsoft.com/contribute/dotnet/dotnet-contribute-code-analysis#contribute-docs-for-caxxxx-rules
- https://github.com/dotnet/roslyn-analyzers/blob/main/GuidelinesForNewRules.md

If your Pull Request is doing one of the following:

- Adding a new diagnostic analyzer or a code fix
- Adding or updating resource strings used by analyzers and code fixes
- Updating analyzer package versions in [Versions.props](../eng/Versions.props)

Then, make sure to run `msbuild /t:pack /v:m` in the repository root; otherwise, the CI build will fail.

- This command must be run from a Visual Studio Developer Command Prompt
- Alternatively, `dotnet msbuild RoslynAnalyzers.sln -t:pack -v:m` can be used from a standard terminal window

Note: Consider merging the PR base branch (`2.9.x`, `main`, or `release/*`) into your branch before you run the pack command to reduce merge conflicts.

-->
This PR adds an analyzer which flags usages of an `.Any()` call which could be replaced with a ```Count```, ```Length``` or ```IsEmpty``` check.

My VB tests are currently failing due to an indentation issue. Maybe someone more versed in VB than me could take a look and tell me what I am missing?

Fixes dotnet/runtime#75933

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,140 @@
+            IsEmptyId,
+            CreateLocalizableResourceString(nameof(PreferIsEmptyOverAnyTitle)),
+            CreateLocalizableResourceString(nameof(PreferIsEmptyOverAnyMessage)),
+            "Performance",
```
```suggestion
            DiagnosticCategory.Performance,
```

CollinAlpert replied:
Done.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,140 @@
+            "Performance",
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(PreferIsEmptyOverAnyDescription)),
+            false,
+            false
```
I prefer named arguments here.

CollinAlpert replied:
Done.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,140 @@
+            LengthId,
+            CreateLocalizableResourceString(nameof(PreferLengthOverAnyTitle)),
+            CreateLocalizableResourceString(nameof(PreferLengthOverAnyMessage)),
+            "Performance",
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(PreferLengthOverAnyDescription)),
+            false,
+            false
```
Same suggestions

CollinAlpert replied:
Done.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,140 @@
+            CountId,
+            CreateLocalizableResourceString(nameof(PreferCountOverAnyTitle)),
+            CreateLocalizableResourceString(nameof(PreferCountOverAnyMessage)),
+            "Performance",
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(PreferCountOverAnyDescription)),
+            false,
+            false
```
Same suggestions

CollinAlpert replied:
Done.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,140 @@
+            if (IsEligibleAnyMethod(originalMethod))
+            {
+                // The collection will be passed as the instance in VB and as an argument in C#. 
+                var firstArgument = invocation.Instance ?? invocation.Arguments[0].Value;
```
Can this crash if the invocation is `System.Linq.Enumerable.Any(/*no argument passed*/)`?

Youssef1313 replied:
I'm talking about this error scenario:

```csharp
public class C
{
    public void M()
    {
        System.Linq.Enumerable.Any();
    }
}
```

I don't know if roslyn generates an invalid operation or invocation operation in this case, but good to test IMO just to confirm.

Youssef1313 replied:
Also consider testing this:

```csharp
public class C
{
    public void M()
    {
        System.Linq.Enumerable.Any<string>(null);
    }
}
```

CollinAlpert replied:
Will do :thumbsup:

CollinAlpert replied:
@Youssef1313 I have the following test now:
```csharp
        [Fact]
        public Task DontWarnOnInvalidCallAsync()
        {
            const string code = @"
public bool HasAny()
{
    return System.Linq.Enumerable.Any();
}";

            return VerifyCS.VerifyAnalyzerAsync(string.Format(CSharpTemplate, code));
        }
```
However, since it is invalid code, the test fails immediately. Is this what you wanted to test?

<img width="758" alt="Screenshot 2022-11-05 at 16 05 39" src="https://user-images.githubusercontent.com/14217185/200126419-a85c2fea-afe5-4ee6-9ed3-f5f411f04a0d.png">

Youssef1313 replied:
Yes, that's what I had in mind to test.

You'll need to adjust it to the following for it to pass:

```csharp
        [Fact]
        public Task DontWarnOnInvalidCallAsync()
        {
            const string code = @"
public bool HasAny()
{
    return System.Linq.Enumerable.{|CS1501:Any|}();
}";

            return VerifyCS.VerifyAnalyzerAsync(string.Format(CSharpTemplate, code));
        }

```

CollinAlpert replied:
TIL, thanks!

CollinAlpert replied:
Done.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,140 @@
+            }
+
+            return typeSymbol.GetMembers(LengthText)
+                .Any(symbol => symbol is IPropertySymbol property && property.Type.SpecialType == SpecialType.System_Int32);
```
`uint` is common for Length/Count in UWP, e.g:

https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.data.itemindexrange.length?view=winrt-18362

I'm not sure if we should handle it too.

@stephentoub What do you think?

Youssef1313 replied:
Ping @stephentoub @buyaa-n for this question.

buyaa-n replied:
Makes sense to cover

Youssef1313 commented on src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Performance/CSharpPreferLengthCountIsEmptyOverAny.Fixer.cs:
```diff
@@ -0,0 +1,89 @@
+    [ExportCodeFixProvider(LanguageNames.CSharp), Shared]
+    public sealed class CSharpPreferLengthCountIsEmptyOverAnyFixer : PreferLengthCountIsEmptyOverAnyFixer
+    {
+        protected override SyntaxNode? ReplaceAnyWithIsEmpty(SyntaxNode root, SyntaxNode node)
```
Does this handle the niche case where `Any` is invoked as an ordinary method with `System.Linq.Enumerable.Any(collection)`?

paulomorgado replied:
I don't know, but I recall working on an analyzer like this a couple of years ago and the intention was to not catch and arbitrary `Any` extension method.

CollinAlpert replied:
Yeah, I can see this becoming problematic. Maybe it is better just to handle LINQ's `Any` method?

CollinAlpert replied:
I added tests for invocation as a normal static method instead of as an extension method.

Youssef1313 commented on src/NetAnalyzers/VisualBasic/Microsoft.NetCore.Analyzers/Performance/BasicPreferLengthCountIsEmptyOverAnyFixer.vb:
```diff
@@ -0,0 +1,88 @@
+        Inherits PreferLengthCountIsEmptyOverAnyFixer
+
+        Protected Overrides Function ReplaceAnyWithIsEmpty(root As SyntaxNode, node As SyntaxNode) As SyntaxNode
+            Dim invocation = TryCast(node, InvocationExpressionSyntax)
```
Parentheses aren't required for method calls in VB. In that case, we won't have InvocationExpressionSyntax

CollinAlpert replied:
Done.

CollinAlpert commented:
I ran the `./build.sh` script and also `dotnet msbuild -t:pack -v:m` (both succeed), however the `AnalyzerReleases.Unshipped.md` file is not being populated. Thus, the CI build fails. Am I missing something?

buyaa-n commented:
> I ran the ./build.sh script and also dotnet msbuild -t:pack -v:m (both succeed), however the AnalyzerReleases.Unshipped.md file is not being populated. Thus, the CI build fails. Am I missing something?

Ah, that is not added with `dotnet msbuild -t:pack -v:m`, it should be added by the code fixer of the `RS2000` analyzer, if you are using VS it looks like this, I think you can add that row manually too
![image](https://user-images.githubusercontent.com/45579687/210671546-026ff0bd-f9f1-41eb-b520-a4ca52c2cc5a.png)

CollinAlpert commented:
@buyaa-n Thanks for the hint! I am using Rider, so I ran `dotnet format analyzers --diagnostics RS2000`.

buyaa-n commented on src/NetAnalyzers/VisualBasic/Microsoft.NetCore.Analyzers/Performance/BasicPreferLengthCountIsEmptyOverAnyFixer.vb:
```diff
@@ -0,0 +1,134 @@
+Imports System.Composition
```
```suggestion
' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.

Imports System.Composition
```

buyaa-n replied:
@mavasani @jmarolf when VB code missing the copyright header the VB build fails without any error/warning info locally. On CI the build never stops, not sure if it is known issue or it was expected

Youssef1313 replied:
@buyaa-n Consider opening an issue to dotnet/roslyn with a commit hash that reproduces the issue

buyaa-n replied:
I am not sure it is roslyn issue, so [created an issue in this repo](https://github.com/dotnet/roslyn-analyzers/issues/6417), we can transfer if needed

Youssef1313 replied:
For some reason, I can't reproduce the same behavior.

though I reproduce a different seemingly buggy behavior. For me, locally, removing the file header causes the build to pass completely, only in the IDE there is a warning.

![image](https://user-images.githubusercontent.com/31348972/211090357-87abcf92-22e6-4551-9de3-0c49332f4883.png)

CollinAlpert commented:
Looks like only my VB tests are failing now. Although I am still clueless as to where that indentation issue is coming from.

buyaa-n commented:
> Looks like only my VB tests are failing now. Although I am still clueless as to where that indentation issue is coming from.

Can't you repro it locally? I can, somehow fixer produced source has an extra spacing before `End Function`, though it might not caused by your code:
```log
    Context: Iterative code fix application
content of '/0/Test0.vb' did not match. Diff shown with expected as baseline:
 
 Imports System.Collections.Immutable
 Imports System.Linq
 
 Public Class Tests
     
 Public Function HasContents(array As ImmutableArray(Of Integer)) As Boolean
     Return Not array.IsEmpty
-End Function
+    End Function
 End Class
```
Looks like VB fixer issue, @mavasani @jmarolf is this a known issue? Is there any work around except adjusting the spacing?

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,142 @@
+
+        private static bool IsEligibleAnyMethod(IMethodSymbol method)
+        {
+            return method.Name == AnyText
+                   && method.ReturnType.SpecialType == SpecialType.System_Boolean
+                   && method.Language == LanguageNames.CSharp && method.Parameters.Length == 1
+                   || (method.Language == LanguageNames.VisualBasic && (method.Parameters.Length == 0 || method.Parameters.Length == 1 && method.Parameters[0].Name == SourceParameterName));
```
Something looks smelly here. In VB, you don't check the method name and return type at all.
In C#, you don't check the parameter name.

In general, I think this should be a symbol comparison.

buyaa-n replied:
> In general, I think this should be a symbol comparison.

In case we want to restrict to a specific symbol like `Enumerable.Any()`

buyaa-n replied:
Turns out we want to flag for any type, not only IEnumerable.Any()

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,142 @@
+            if (IsEligibleAnyMethod(originalMethod))
+            {
+                // The collection will be passed as the instance in VB and as an argument in C#. 
+                var firstArgument = invocation.Instance ?? invocation.Arguments[0].Value;
+                var type = (firstArgument as IConversionOperation)?.Operand.Type ?? firstArgument.Type;
```
@mavasani Is this logic correct? The logic in `GetReceiverType` seems more expensive (involves semantic model calls).

for now, I think this should use `GetReceiverType`, and if this logic is correct and more efficient, then `GetReceiverType` should be updated as a follow up. What do you think?

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,142 @@
+using System.Collections;
+using System.Collections.Immutable;
+using System.Linq;
+using Analyzer.Utilities;
```
```suggestion
using Analyzer.Utilities;
using Analyzer.Utilities.Extensions;
```

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,142 @@
+        internal const string LengthId = "CA1861";
+        internal const string CountId = "CA1862";
+
+        private const string SourceParameterName = "source";
+
```
```suggestion
```

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,142 @@
+            isPortedFxCopRule: false,
+            isDataflowRule: false
+        );
+
```
```suggestion

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(
            LengthDescriptor,
            CountDescriptor,
            IsEmptyDescriptor
        );

```

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,142 @@
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterOperationAction(OnInvocationAnalysis, OperationKind.Invocation);
```
```suggestion
            context.RegisterCompilationStartAction(context =>
            {
                var anyMethod = WellKnownTypeProvider.GetOrCreate(context.Compilation).GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemLinqEnumerable)
                    ?.GetMembers(AnyText).OfType<IMethodSymbol>().FirstOrDefault(m => m.IsExtensionMethod && m.Parameters.Length == 1);
                if (anyMethod is not null)
                {
                    context.RegisterOperationAction(context => OnInvocationAnalysis(context, anyMethod), OperationKind.Invocation);
                }
            });
```

CollinAlpert replied:
The problem with this approach is that we miss things like https://learn.microsoft.com/en-us/dotnet/api/system.linq.immutablearrayextensions.any?view=net-7.0#system-linq-immutablearrayextensions-any-1(system-collections-immutable-immutablearray((-0)))

Youssef1313 replied:
@CollinAlpert I see. Then consider not adding anything to compilation start, but check the method correctly.

The code should be language-agnostic (no C#/VB checks), and parameter name shouldn't be checked. Just check method name, whether it's extension method, and whether it has a single parameter.

CollinAlpert replied:
I remember having problems with eliminating language checks, as I believe Roslyn considers `Any` to be an extension method in C# and not an extension method in VB.

Youssef1313 replied:
@CollinAlpert That suggestion to use `ReducedFrom` if `MethodKind` is `ReducedExtension` should probably fix the problems you were seeing.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,142 @@
+            context.RegisterOperationAction(OnInvocationAnalysis, OperationKind.Invocation);
+        }
+
+        private static void OnInvocationAnalysis(OperationAnalysisContext context)
```
```suggestion
        private static void OnInvocationAnalysis(OperationAnalysisContext context, IMethodSymbol anyMethod)
```

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,142 @@
+        private static void OnInvocationAnalysis(OperationAnalysisContext context)
+        {
+            var invocation = (IInvocationOperation)context.Operation;
+            var originalMethod = invocation.TargetMethod.OriginalDefinition;
```
```suggestion
            var originalMethod = invocation.TargetMethod.OriginalDefinition;
            if (originalMethod.MethodKind == MethodKind.ReducedExtension)
            {
                originalMethod = originalMethod.ReducedFrom;
            }

```

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,142 @@
+        {
+            var invocation = (IInvocationOperation)context.Operation;
+            var originalMethod = invocation.TargetMethod.OriginalDefinition;
+            if (IsEligibleAnyMethod(originalMethod))
```
```suggestion
            if (originalMethod.Equals(anyMethod, SymbolEqualityComparer.Default))
```

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,142 @@
+            var originalMethod = invocation.TargetMethod.OriginalDefinition;
+            if (IsEligibleAnyMethod(originalMethod))
+            {
+                // The collection will be passed as the instance in VB and as an argument in C#. 
+                var firstArgument = invocation.Instance ?? invocation.Arguments[0].Value;
+                var type = (firstArgument as IConversionOperation)?.Operand.Type ?? firstArgument.Type;
```
```suggestion
                var type = invocation.GetReceiverType(context.Compilation, beforeConversion: true, context.CancellationToken);
                if (type is null)
                {
                    return;
                }
```

CollinAlpert replied:
Is there a way to make `GetReceiverType` work for arrays as well? Right now it returns `null` when `Any` is called on an array.

Youssef1313 replied:
That looks like a bug in `GetReceiverType`.

I think the fix would be changing the return type to `ITypeSymbol` and change all `as INamedTypeSymbol` to `ITypeSymbol`

CollinAlpert replied:
Indeed it does. However it might be better to just manually check if it is an array at the call site when `GetReceiverType` returns `null`.

Youssef1313 replied:
> However it might be better to just manually check if it is an array at the call site when GetReceiverType returns null.

My feeling is that this should be fixed in `GetReceiverType`. @mavasani thoughts?

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,142 @@
+                var type = (firstArgument as IConversionOperation)?.Operand.Type ?? firstArgument.Type;
+                if (HasEligibleIsEmptyProperty(type))
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(IsEmptyDescriptor, invocation.Syntax.GetLocation()));
```
```suggestion
                    context.ReportDiagnostic(invocation.CreateDiagnostic(IsEmptyDescriptor));
```

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,142 @@
+
+                if (HasEligibleLengthProperty(type))
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(LengthDescriptor, invocation.Syntax.GetLocation()));
```
```suggestion
                    context.ReportDiagnostic(invocation.CreateDiagnostic(LengthDescriptor));
```

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,142 @@
+
+                if (HasEligibleCountProperty(type))
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(CountDescriptor, invocation.Syntax.GetLocation()));
```
```suggestion
                    context.ReportDiagnostic(invocation.CreateDiagnostic(CountDescriptor));
```

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,142 @@
+                }
+            }
+        }
+
+        private static bool IsEligibleAnyMethod(IMethodSymbol method)
+        {
+            return method.Name == AnyText
+                   && method.ReturnType.SpecialType == SpecialType.System_Boolean
+                   && method.Language == LanguageNames.CSharp && method.Parameters.Length == 1
+                   || (method.Language == LanguageNames.VisualBasic && (method.Parameters.Length == 0 || method.Parameters.Length == 1 && method.Parameters[0].Name == SourceParameterName));
+        }
```
```suggestion
```

Youssef1313 replied:
Given that we need to catch any `Any` method. The check should be:

```csharp
return method is
    {
        Name: AnyText,
        ReturnType.SpecialType: SpecialType.System_Boolean,
        IsExtensionMethod: true, 
        Parameters: [_]
    };
```

On the callsite, the passed method should be:

```csharp
var targetMethod = invocation.TargetMethod;
if (targetMethod.MethodKind == MethodKind.ReducedExtension)
{
    targetMethod = targetMethod.ReducedFrom;
}

// pass targetMethod to IsEligibleAnyMethod at this point.
```

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,142 @@
+            return typeSymbol.GetMembers(CountText)
+                .Any(symbol => symbol is IPropertySymbol property && property.Type.SpecialType == SpecialType.System_Int32);
+        }
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(
+            LengthDescriptor,
+            CountDescriptor,
+            IsEmptyDescriptor
+        );
```
```suggestion
```

Youssef1313 commented:
> Looks like VB fixer issue, @mavasani @jmarolf is this a known issue? Is there any work around except adjusting the spacing?

@buyaa-n To me it looks like this is a test authoring issue. The input source code isn't formatted correctly.

Youssef1313 commented on src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Performance/CSharpPreferLengthCountIsEmptyOverAny.Fixer.cs:
```diff
@@ -0,0 +1,97 @@
+                expression,
+                IdentifierName(PreferLengthCountIsEmptyOverAnyAnalyzer.IsEmptyText)
+            );
+            if (invocation.Parent is PrefixUnaryExpressionSyntax prefixExpression && prefixExpression.IsKind(SyntaxKind.LogicalNotExpression))
```
```suggestion
            if (invocation.Parent.IsKind(SyntaxKind.LogicalNotExpression))
```

Youssef1313 commented on src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Performance/CSharpPreferLengthCountIsEmptyOverAny.Fixer.cs:
```diff
@@ -0,0 +1,97 @@
+            );
+            if (invocation.Parent is PrefixUnaryExpressionSyntax prefixExpression && prefixExpression.IsKind(SyntaxKind.LogicalNotExpression))
+            {
+                return root.ReplaceNode(prefixExpression, newMemberAccess.WithTriviaFrom(prefixExpression));
```
```suggestion
                return root.ReplaceNode(invocation.Parent, newMemberAccess.WithTriviaFrom(invocation.Parent));
```

Youssef1313 commented on src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Performance/CSharpPreferLengthCountIsEmptyOverAny.Fixer.cs:
```diff
@@ -0,0 +1,97 @@
+                );
+            }
+
+            if (invocation.Parent is PrefixUnaryExpressionSyntax prefixExpression && prefixExpression.IsKind(SyntaxKind.LogicalNotExpression))
```
```suggestion
            if (invocation.Parent.IsKind(SyntaxKind.LogicalNotExpression))
```

Youssef1313 commented on src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Performance/CSharpPreferLengthCountIsEmptyOverAny.Fixer.cs:
```diff
@@ -0,0 +1,97 @@
+            {
+                var binaryExpression = GetBinaryExpression(expression, propertyName, SyntaxKind.EqualsExpression);
+
+                return root.ReplaceNode(prefixExpression, binaryExpression.WithTriviaFrom(prefixExpression));
```
```suggestion
                return root.ReplaceNode(invocation.Parent, binaryExpression.WithTriviaFrom(invocation.Parent));
```

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAny.Fixer.cs:
```diff
@@ -0,0 +1,54 @@
+            {
+                var (newRoot, codeFixTitle) = diagnostic.Id switch
+                {
+                    PreferLengthCountIsEmptyOverAnyAnalyzer.IsEmptyId => (ReplaceAnyWithIsEmpty(root!, node), MicrosoftNetCoreAnalyzersResources.PreferIsEmptyOverAnyCodeFixTitle),
```
@mavasani Is it okay to calculate the whole replacement in `RegisterCodeFixesAsync` directly outside of `CodeAction`?

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAny.Fixer.cs:
```diff
@@ -0,0 +1,54 @@
+
+        public override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+        public override ImmutableArray<string> FixableDiagnosticIds { get; } = ImmutableArray.Create(
+            PreferLengthCountIsEmptyOverAnyAnalyzer.LengthId,
+            PreferLengthCountIsEmptyOverAnyAnalyzer.CountId,
+            PreferLengthCountIsEmptyOverAnyAnalyzer.IsEmptyId
+        );
```
Move before `RegisterCodeFixesAsync`

Youssef1313 commented on src/NetAnalyzers/VisualBasic/Microsoft.NetCore.Analyzers/Performance/BasicPreferLengthCountIsEmptyOverAnyFixer.vb:
```diff
@@ -0,0 +1,136 @@
+                    Return Nothing
+                End If
+
+                Dim unaryParent = TryCast(memberAccess.Parent, UnaryExpressionSyntax)
+                If unaryParent IsNot Nothing And unaryParent.IsKind(SyntaxKind.NotExpression) Then
+                    Dim binaryExpression = GetBinaryExpression(memberAccess.Expression, propertyName, SyntaxKind.EqualsExpression)
+
+                    Return root.ReplaceNode(unaryParent, binaryExpression.WithTriviaFrom(unaryParent))
+                End If
```
```suggestion
                If memberAccess.Parent.IsKind(SyntaxKind.NotExpression) Then
                    Dim binaryExpression = GetBinaryExpression(memberAccess.Expression, propertyName, SyntaxKind.EqualsExpression)

                    Return root.ReplaceNode(memberAccess.Parent, binaryExpression.WithTriviaFrom(memberAccess.Parent))
                End If
```

Youssef1313 commented on src/NetAnalyzers/VisualBasic/Microsoft.NetCore.Analyzers/Performance/BasicPreferLengthCountIsEmptyOverAnyFixer.vb:
```diff
@@ -0,0 +1,136 @@
+                End If
+
+                Dim expression = memberAccess.Expression
+                If invocation?.ArgumentList.Arguments.Count > 0 Then
```
```suggestion
                If invocation.ArgumentList.Arguments.Count > 0 Then
```

Youssef1313 commented on src/NetAnalyzers/VisualBasic/Microsoft.NetCore.Analyzers/Performance/BasicPreferLengthCountIsEmptyOverAnyFixer.vb:
```diff
@@ -0,0 +1,136 @@
+                    expression = invocation.ArgumentList.Arguments(0).GetExpression()
+                End If
+
+                Dim unaryParent = TryCast(invocation.Parent, UnaryExpressionSyntax)
+                If unaryParent IsNot Nothing And unaryParent.IsKind(SyntaxKind.NotExpression) Then
+                    Dim binaryExpression = GetBinaryExpression(expression, propertyName, SyntaxKind.EqualsExpression)
+
+                    Return root.ReplaceNode(unaryParent, binaryExpression.WithTriviaFrom(unaryParent))
+                End If
+
+                Return root.ReplaceNode(invocation, GetBinaryExpression(expression, propertyName, SyntaxKind.NotEqualsExpression).WithTriviaFrom(invocation))
```
```suggestion
                If invocation.Parent.IsKind(SyntaxKind.NotExpression) Then
                    Dim binaryExpression = GetBinaryExpression(expression, propertyName, SyntaxKind.EqualsExpression)

                    Return root.ReplaceNode(invocation.Parent, binaryExpression.WithTriviaFrom(invocation.Parent))
                End If

                Return root.ReplaceNode(invocation, GetBinaryExpression(expression, propertyName, SyntaxKind.NotEqualsExpression).WithTriviaFrom(invocation))
```

Youssef1313 reviewed (changes requested):
I reviewed the analyzer in detail and added few comments (I think all are ready-to-commit suggestions)

Didn't have enough time to continue reviewing the codefix in details, but got few comments.

Youssef1313 commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/PreferCountOverAnyTests.cs:
```diff
@@ -0,0 +1,286 @@
+{
+    public class PreferCountOverAnyTests
+    {
+        private const string CSharpTemplate = @"
+using System.Collections.Generic;
+using System.Linq;
+
+public class Tests {{
+    {0}
+}}";
+
+        private const string VbTemplate = @"
+Imports System.Collections.Generic
+Imports System.Linq
+
+Public Class Tests
+    {0}
+End Class";
```
I personally prefer to avoid these.

CollinAlpert replied:
What would be a preferred approach? Should I put these in each test method?

Youssef1313 replied:
IMO, yes. I don't see the duplication as a problem.

buyaa-n commented:
> @buyaa-n To me it looks like this is a test authoring issue. The input source code isn't formatted correctly.

Ah you are right, the template and input string has formatting issue

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1998,4 +1998,40 @@
+    <value>Prefer 'Length' over 'Any()'</value>
+  </data>
+  <data name="PreferLengthOverAnyMessage" xml:space="preserve">
+    <value>Prefer a 'Length' check over 'Any()' to avoid the need for devirtualization</value>
```
NIT:  For what meaning the `devirtualization` is used? It seems used for handling virtual functions
```suggestion
    <value>Prefer a 'Length' check over 'Any()' to avoid unnecessary enumeration</value>
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1998,4 +1998,40 @@
+    <value>Prefer a 'Length' check over an 'Any()' call. The 'Any()' call will convert the collection into an IEnumerable which creates the need for devirtualization at runtime.</value>
+  </data>
+  <data name="PreferCountOverAnyCodeFixTitle" xml:space="preserve">
+    <value>Prefer 'Count' over 'Any()'</value>
```
```suggestion
    <value>Use 'Count' check instead of 'Any()'</value>
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1998,4 +1998,40 @@
+    <value>Prefer a 'Count' check over an 'Any()' call. The 'Any()' call will convert the collection into an IEnumerable which creates the need for devirtualization at runtime.</value>
+  </data>
+  <data name="PreferIsEmptyOverAnyCodeFixTitle" xml:space="preserve">
+    <value>Prefer 'IsEmpty' over 'Any()'</value>
```
```suggestion
    <value>Use 'IsEmpty' check instead of 'Any()'</value>
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1998,4 +1998,40 @@
     <value>Change type of parameter '{0}' from '{1}' to '{2}' for improved performance</value>
   </data>
+  <data name="PreferLengthOverAnyCodeFixTitle" xml:space="preserve">
+    <value>Prefer 'Length' over 'Any()'</value>
```
```suggestion
    <value>Use 'Length' check instead of 'Any()'</value>
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1998,4 +1998,40 @@
+    <value>Prefer 'IsEmpty' over 'Any()'</value>
+  </data>
+  <data name="PreferIsEmptyOverAnyTitle" xml:space="preserve">
+    <value>Prefer 'IsEmpty' over 'Any()'</value>
```
```suggestion
    <value>Prefer 'IsEmpty' check over 'Any()'</value>
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1998,4 +1998,40 @@
+    <value>Prefer 'Length' over 'Any()'</value>
+  </data>
+  <data name="PreferLengthOverAnyTitle" xml:space="preserve">
+    <value>Prefer 'Length' over 'Any()'</value>
```
Here and below, I think better to add `check`
```suggestion
    <value>Prefer 'Length' check over 'Any()'</value>
```

buyaa-n commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/PreferIsEmptyOverAnyTests.cs:
```diff
@@ -0,0 +1,260 @@
+        {
+            const string code = @"
+public bool HasContents(ImmutableArray<int> array) {
+    return {|#0:System.Linq.Enumerable.Any(array)|};
```
ImmutableArray.Any() just checks [`immutableArray.Length > 0;`](https://source.dot.net/#System.Collections.Immutable/System/Linq/ImmutableArrayExtensions.cs,99), so probably we should not warn on this. I wonder if we should flag only if the IEnumerable.Any() extension method is used CC @stephentoub

buyaa-n replied:
Got a response that we want to keep flagging it, if not for perf then for clarity.

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,142 @@
+namespace Microsoft.NetCore.Analyzers.Performance
+{
+    /// <summary>
+    /// Prefer .Length/Count/IsEmpty over Any()
```
```suggestion
    /// Prefer .Length/Count/IsEmpty check over Any()
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1998,4 +1998,40 @@
+    <value>Prefer 'Count' over 'Any()'</value>
+  </data>
+  <data name="PreferCountOverAnyTitle" xml:space="preserve">
+    <value>Prefer 'Count' over 'Any()'</value>
```
Probably add property to make it less confusing with[ `DoNotUseCountWhenAnyCanBeUsed`](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1827)
```suggestion
    <value>Prefer 'Count' property check over 'Any()'</value>
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,142 @@
+
+        private static bool HasEligibleIsEmptyProperty(ITypeSymbol typeSymbol)
+        {
+            return typeSymbol.GetMembers(IsEmptyText)
+                .Any(symbol => symbol is IPropertySymbol property && property.Type.SpecialType == SpecialType.System_Boolean);
```
```suggestion
            return typeSymbol.GetMembers(IsEmptyText).OfType<IPropertySymbol>()
                .Any(property => property.Type.SpecialType == SpecialType.System_Boolean);
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,142 @@
+                return true;
+            }
+
+            return typeSymbol.GetMembers(LengthText)
+                .Any(symbol => symbol is IPropertySymbol property && property.Type.SpecialType == SpecialType.System_Int32);
```
```suggestion
            return ypeSymbol.GetMembers(CountText).OfType<IPropertySymbol>()
                .Any(property => property.Type.SpecialType is SpecialType.System_Int32 or SpecialType.System_UInt32);
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,142 @@
+
+        private static bool HasEligibleCountProperty(ITypeSymbol typeSymbol)
+        {
+            return typeSymbol.GetMembers(CountText)
+                .Any(symbol => symbol is IPropertySymbol property && property.Type.SpecialType == SpecialType.System_Int32);
```
```suggestion
            return typeSymbol.GetMembers(CountText).OfType<IPropertySymbol>()
                .Any(property => property.Type.SpecialType == SpecialType.System_Int32);
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1998,4 +1998,40 @@
+    <value>Prefer 'Count' over 'Any()'</value>
+  </data>
+  <data name="PreferCountOverAnyMessage" xml:space="preserve">
+    <value>Prefer a 'Count' check over 'Any()' to avoid the need for devirtualization</value>
```
```suggestion
    <value>Prefer comparing 'Count' to 0 rather than using 'Any()', both for clarity and for performance</value>
```

buyaa-n replied:
Please update the description and other messages accordingly

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAny.Fixer.cs:
```diff
@@ -0,0 +1,54 @@
+                    PreferLengthCountIsEmptyOverAnyAnalyzer.IsEmptyId => (ReplaceAnyWithIsEmpty(root!, node), MicrosoftNetCoreAnalyzersResources.PreferIsEmptyOverAnyCodeFixTitle),
+                    PreferLengthCountIsEmptyOverAnyAnalyzer.LengthId => (ReplaceAnyWithLength(root!, node), MicrosoftNetCoreAnalyzersResources.PreferLengthOverAnyCodeFixTitle),
+                    PreferLengthCountIsEmptyOverAnyAnalyzer.CountId => (ReplaceAnyWithCount(root!, node), MicrosoftNetCoreAnalyzersResources.PreferCountOverAnyCodeFixTitle),
+                    _ => throw new ArgumentOutOfRangeException(nameof(context))
```
```suggestion
                    _ => throw new NotSupportedException()
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,142 @@
+        internal const string LengthText = nameof(Array.Length);
+        internal const string CountText = nameof(ICollection.Count);
+
+        internal const string IsEmptyId = "CA1860";
+        internal const string LengthId = "CA1861";
+        internal const string CountId = "CA1862";
```
The diagnostics quite similar, do not see need to handle them separately, probably better to be one diagnostic?
CC @stephentoub @mavasani

buyaa-n replied:
Resolving this comment as:
1. Other similar analyzers using different ids
2. Personally in case it was not `IEnumerable.Any()` that does count/length check in the implementation I would prefer to use Any() over Count or Length check, therefore it seems one might prefer to suppress one of them over the other.
3. With different diagnostics title and description message are cleaner

buyaa-n replied:
I am unresolving this as if we restrict to `Enumerable.Any()` only there will be no need to handle these warnings separately. 

@CollinAlpert lets use only one ID for each DiagnosticDecriptors

buyaa-n commented:
@CollinAlpert [found one false positive case](https://github.com/dotnet/runtime/blob/main/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/FlowAnnotations.cs#L595) in runtime test:
```cs
if (annotatedFields.Any(a => a.Field == backingField)) // flags this
{
    _logger.LogWarning(backingField, DiagnosticId.DynamicallyAccessedMembersOnPropertyConflictsWithBackingField, property.GetDisplayName(), backingField.GetDisplayName());
}
```
overall, the PR looks quite close to complete. Please apply the feedbacks and fix failed VB tests

Youssef1313 commented:
```csharp
using System.Collections;
using System.Collections.Generic;
using System.Linq;

class C : IEnumerable<int>
{
    // Converting Any() to Count changes behavior and goes into a loop.
    public int Count => !this.Any() ? 0 : this.Count();

    public IEnumerator<int> GetEnumerator() => null!;

    IEnumerator IEnumerable.GetEnumerator() => null!;

}
```

I'm not sure if this is a realistic scenario and/or if we want to handle it.

CollinAlpert commented:
Thank you so much for your input! I have addressed all comments and implemented the requested changes. The only thing still missing are a few failing tests due to a bug in `GetReceiverType` (see https://github.com/dotnet/roslyn-analyzers/pull/6236#discussion_r1064012974).

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,155 @@
+        {
+            return typeSymbol.GetMembers(IsEmptyText)
+                .OfType<IPropertySymbol>()
+                .Any(property => property.Type.SpecialType == SpecialType.System_Boolean);
```
```suggestion
                .Any(property => property.Type.SpecialType is SpecialType.System_Int32 or SpecialType.System_UInt32);
```

buyaa-n replied:
This `IsEmpty`, should be boolean

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1998,4 +1998,40 @@
+    <value>Prefer comparing 'Length' to 0 rather than using 'Any()', both for clarity and for performance</value>
+  </data>
+  <data name="PreferLengthOverAnyDescription" xml:space="preserve">
+    <value>Prefer a 'Length' check over an 'Any()' call. The 'Any()' call will convert the collection into an 'IEnumerable' which creates the need for devirtualization at runtime.</value>
```
```suggestion
    <value>Prefer comparing 'Length' to 0 rather than using 'Any()'. It is clearer and the 'Any()' call could be slower especially if it uses the 'IEnumerable.Any()' extension method</value>
```
The wording still could use more improvement, CC @gewarren @stephentoub

gewarren commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1998,4 +1998,40 @@
+    <value>Prefer comparing 'Length' to 0 rather than using 'Any()', both for clarity and for performance</value>
+  </data>
+  <data name="PreferLengthOverAnyDescription" xml:space="preserve">
+    <value>Prefer comparing 'Length' to 0 rather than using 'Any()'. It is clearer and the 'Any()' call could be slower especially if it uses the 'IEnumerable.Any()' extension method.</value>
```
```suggestion
    <value>Prefer comparing 'Length' to 0 rather than calling 'Any()'. The intent is clearer and the 'Any()' call could be slower, especially if it uses the 'IEnumerable.Any()' extension method.</value>
```

CollinAlpert replied:
Done. Thanks for your input!

CollinAlpert commented:
From what I can tell, the only thing is missing is to make a decision on the bug in `GetReceiverType`. @Youssef1313 what do you think would be the best way to proceed here?

Youssef1313 commented:
@CollinAlpert I think you can just change the return type from `INamedTypeSymbol` to `ITypeSymbol` so that this method could return array types.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAny.Fixer.cs:
```diff
@@ -0,0 +1,54 @@
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var node = root?.FindNode(context.Span);
+            if (node is null)
+            {
+                return;
+            }
```
I think the `Microsoft.CodeAnalysis` version used in this repo has non-nullable return for `GetSyntaxRootAsync`, so null check is unnecessary.

```suggestion
            var node = root.FindNode(context.Span);
```

Later, when we move to a newer version where `root` will become nullable, it's more likely we'll introduce a `GetRequiredSyntaxRootAsync` extension.

Youssef1313 commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/PreferIsEmptyOverAnyTests.cs:
```diff
@@ -0,0 +1,386 @@
+public class Tests {
+    public void M() {
+        var array = ImmutableArray<int>.Empty;
+        _ = {|#0:array.Any()|};
```
Please add a test where `array.Any()` is passed as an argument to a method that accepts a boolean. If the codefix crashes, you'll need to pass `getInnermostNodeForTie: true` to `FindNode`

buyaa-n reviewed (approved):
Thank you @CollinAlpert LGTM and thank you @Youssef1313 for review! 

I will hold off merging as @jmarolf might want to take a look

jmarolf commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,155 @@
+        {
+            return method is
+            {
+                Name: AnyText,
```
This diagnostic is about _specific_ implementations of `Count()` and `Any()`. We only know this is true for the one that come from the framework. If there are other extension methods that _happen_ to have these names we would have no idea if this is a correct thing to suggest. I would like us to also match on the namespace that the extension methods come from to reduce the number of false positives.

stephentoub replied:
Yes, this should specifically be about System.Linq.Enumerable.Count and System.Linq.Enumerable.Any.

buyaa-n replied:
I had [similar comment about this](https://github.com/dotnet/roslyn-analyzers/pull/6236#discussion_r1063840526) and had offline discussion and my understanding was we prefer to apply the rule for Any() method of other types too. Though personally I prefer to restrict it to `Enumerable.Any`. 

@CollinAlpert please [do a symbol check](https://github.com/dotnet/roslyn-analyzers/pull/6236#discussion_r1063703944) instead as @Youssef1313 suggested before

stephentoub replied:
> my understanding was we prefer to apply the rule for Any() method of other types too

If we know what those methods are because we own them, I think that's fine, e.g. you had asked me offline about ImmutableArrayExtensions.Any, which is an extension defined in System.Collections.Immutable, i.e. we own it.  I'm fine if we want to special-case that in addition to System.Linq.Enumerable.Any.  But the biggest impact here and the simplest thing to do is focus on just LINQ's Enumerable.Any.

jmarolf replied:
If there are multiple `Any` methods that are defined in dotnet/runtime that we want to check for, I think we should check for those specific methods. I don't expect the list of `Any` extension methods in the framework will be large enough for the list to be hard to maintain.

buyaa-n replied:
> I think that's fine, e.g. you had asked me offline about ImmutableArrayExtensions.Any, which is an extension defined in System.Collections.Immutable, i.e. we own it. 

Sorry, I misunderstood, I took `ImmutableArrayExtensions.Any` as an example for any other `Any()` implementation 

> I'm fine if we want to special-case that in addition to System.Linq.Enumerable.Any. But the biggest impact here and the simplest thing to do is focus on just LINQ's Enumerable.Any.

Right, for me we should focus on just `Enumerable.Any`, some people might argue that `ImmutableArrayExtensions.Any()` is a false positive.

jmarolf commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs:
```diff
@@ -0,0 +1,155 @@
+            };
+        }
+
+        private static bool HasEligibleIsEmptyProperty(ITypeSymbol typeSymbol)
```
I agree with the intent of this heuristic. If properties such as `IsEmpty`, `Length`, or `Count` exist on a collection type and they don't represent information about the number of elements in that collection then the type is out of step with most of .NET.

However, as I mentioned, we are _assuming_ that the type is a collection type. We don't actually check for that yet. I would suggest that we verify that the type our `Any` method applies to inherits from either `IEnumerable` or `IEnumerable<T>`.

jmarolf replied:
Consider a record and extension method pair like this:
```csharp
public record UniversityClass(
  Professor? Instructor,
  ImmutableArray<Student> TeachingAssistants,
  ImmutableArray<Student> Students)
{
  // meant to indicate that a class has no students signed up.
  public bool IsEmpty => Students.Length == 0;
}

// meant to indicate that a class has no instructors assigned to teach it.
public static bool Any(UniversityClass @class)
  => @class.Instructor is not null ||
     @class.TeachingAssistants.Length > 0;
```

CollinAlpert commented:
Thanks for the feedback! I have incorporated the changes and now check that the `Any` method is `System.Linq.Enumerable.Any()` and that the corresponding type implements `IEnumerable` or `IEnumerable<T>`.

jmarolf reviewed (approved):
Looks good! thanks for all the work here.

CollinAlpert commented:
Happy to be able to contribute!

buyaa-n reviewed (changes requested):
@CollinAlpert [the different ids](https://github.com/dotnet/roslyn-analyzers/pull/6236/files#r1083043151) not suggested in the API review and I do not see need of using 3 ids, the id ranges are limited, we better reuse same id for all diagnostics. Sorry for requesting it now, in case you are busy, I will try to contribute to the PR with the change

CollinAlpert commented:
No worries, I can change it. However then we need a new message which encapsulates all three properties. Do you have a suggestion?

buyaa-n commented:
> No worries, I can change it. 

Thank you!

> However then we need a new message which encapsulates all three properties.

I think you can create each diagnostic with existing different messages, no?
```cs
        internal const string RuleId = "CA1860";
        internal static readonly DiagnosticDescriptor IsEmptyDescriptor = DiagnosticDescriptorHelper.Create(
            RuleId,
            CreateLocalizableResourceString(nameof(PreferIsEmptyOverAnyTitle)),
            CreateLocalizableResourceString(nameof(PreferIsEmptyOverAnyMessage)),
            DiagnosticCategory.Performance,
            RuleLevel.IdeSuggestion,
            CreateLocalizableResourceString(nameof(PreferIsEmptyOverAnyDescription)),
            isPortedFxCopRule: false,
            isDataflowRule: false
        );

        internal static readonly DiagnosticDescriptor LengthDescriptor = DiagnosticDescriptorHelper.Create(
            RuleId,
            CreateLocalizableResourceString(nameof(PreferLengthOverAnyTitle)),
            CreateLocalizableResourceString(nameof(PreferLengthOverAnyMessage)),
            DiagnosticCategory.Performance,
            RuleLevel.IdeSuggestion,
            CreateLocalizableResourceString(nameof(PreferLengthOverAnyDescription)),
            isPortedFxCopRule: false,
            isDataflowRule: false
        );
       ...
```

buyaa-n commented:
> Though not sure which title/description will be on doc

The generated docs just using the 1st diagnostic description and title, for me that is OK as integrated title and/or description doesn't look good to me. For example: I prefer `Use 'Length' check instead of 'Any()'` over `Use 'Length/ Count/IsEmpty' check instead of 'Any()'`. Is that OK for you @jmarolf @mavasani?

For distinguishing the diagnostics in fixer It seems better to use the properties as checking the DiagnosticDescriptiors somehow not triggering the fixer: https://github.com/dotnet/roslyn-analyzers/blob/9b5d13007e77cc5e5844c2790cd39218a99c9e9b/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseCountProperly.cs#L377

jmarolf commented:
Diagnostic Ids are most important when thinking about how a developer is going to suppress this analyzer. Because I cannot think of a good reason to suppress some of these cases and not others, I agree with @buyaa-n that we should just have once id.
 
 For documentation I think we should have the same Title for all the diagnostics with different descriptions depending on the situation.

buyaa-n commented:
> The generated docs just using the 1st diagnostic description and title, for me that is OK as integrated title and/or description doesn't look good to me.

Hm, it seems the other title/descriptions are not used/visible anywhere else. 

> For documentation I think we should have the same Title for all the diagnostics with different descriptions depending on the situation.

Right, it seems we should use one title, is the different descriptions will be used anywhere?

How is this title sounds? `Avoid using 'Enumerable.Any()' extension method` 

And a description could be something like: `Prefer using from 'IsEmpty', 'Count' or 'Length' properties whichever available, rather than calling ‘Any()'. The intent is clearer and is more performant than 'Enumerable.Any()' extension method.`

cdrfenix commented:
The main discussion thread its closed, but this terrible change, at least is a warning we can disable, but this is favoring performance over legibility.

stephentoub commented:
> The main discussion thread its closed, but this is a bad change, at least its a warning we can disable, but this is favoring performance over legibility.

It's not even a warning, it defaults to info / suggestion. If you're seeing it as a warning, it's because you're doing something to bump up its severity.

(And legibility is in the eye of the beholder; personally, I find explicit use of the properties more readable/understandable.)
